presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +219,143 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            if (childDecorrelationResult.symbolsToPropagate.isEmpty()) {
+                return decorrelatedOrderingScheme
+                        .map(orderingScheme -> Optional.of(new DecorrelationResult(
+                                // ordering symbols are present - return decorrelated TopNNode
+                                new TopNNode(node.getId(), decorrelatedChildNode, node.getCount(), orderingScheme, node.getStep()),
+                                childDecorrelationResult.symbolsToPropagate,
+                                childDecorrelationResult.correlatedPredicates,
+                                childDecorrelationResult.correlatedSymbolsMapping,
+                                node.getCount() == 1)))
+                        .orElseGet(() -> Optional.of(new DecorrelationResult(
+                                // no ordering symbols are left - convert to LimitNode
+                                new LimitNode(node.getId(), decorrelatedChildNode, node.getCount(), false),
+                                childDecorrelationResult.symbolsToPropagate,
+                                childDecorrelationResult.correlatedPredicates,
+                                childDecorrelationResult.correlatedSymbolsMapping,
+                                node.getCount() == 1)));
+            }
+
+            if (!constantSymbols.containsAll(childDecorrelationResult.symbolsToPropagate)) {
+                return Optional.empty();
+            }
+
+            return decorrelatedOrderingScheme
+                    .map(orderingScheme -> {
+                        // ordering symbols are present - rewrite TopN to TopNRowNumberNode partitioned by constant symbols
+                        TopNRowNumberNode topNRowNumberNode = new TopNRowNumberNode(
+                                node.getId(),
+                                decorrelatedChildNode,
+                                new Specification(
+                                        ImmutableList.copyOf(childDecorrelationResult.symbolsToPropagate),
+                                        Optional.of(orderingScheme)),
+                                symbolAllocator.newSymbol("row_number", BIGINT),
+                                toIntExact(node.getCount()),
+                                false,
+                                Optional.empty());
+
+                        return Optional.of(new DecorrelationResult(
+                                topNRowNumberNode,
+                                childDecorrelationResult.symbolsToPropagate,
+                                childDecorrelationResult.correlatedPredicates,
+                                childDecorrelationResult.correlatedSymbolsMapping,
+                                node.getCount() == 1));
+                    })
+                    .orElseGet(() -> {
+                        // no ordering symbols are left - rewrite TopN to RowNumberNode partitioned by constant symbols
```

Comment:
ditto: `TopNRowNumberNode` to `RowNumberNode` feels like it should be separate rule when ordering scheme is empty.

Reply:
`OrderingScheme` cannot be empty.